cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -189,6 +194,73 @@ Status ConstBoolExec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
   return Status::OK();
 }
 
+struct NonZeroVisitor {
+  UInt64Builder* builder;
+  const ArrayData& array;
+
+  NonZeroVisitor(UInt64Builder* builder, const ArrayData& array)
+      : builder(builder), array(array) {}
+
+  Status Visit(const DataType& type) { return Status::NotImplemented(type.ToString()); }
+
+  template <typename Type>
+  enable_if_t<is_primitive_ctype<Type>::value, Status> Visit(const Type&) {
+    using T = typename GetViewType<Type>::T;
+    uint32_t index = 0;
+
+    return VisitArrayDataInline<Type>(
+        this->array,
+        [&](T v) {
+          if (v) {
+            this->builder->UnsafeAppend(index);
+          }
+          ++index;
+          return Status::OK();
+        },
+        [&]() {
+          ++index;
+          return Status::OK();
+        });
+  }
```

Comment:
wouldn't it be best to use the `void` visitor here? AFAIU the hot loop checks if status is okay. But may be the compiler removes that part :thinking: Nevertheless, I think the following would be neater. 
```suggestion
    VisitArrayDataInline<Type>(
        this->array,
        [&](T v) {
          if (v) {
            this->builder->UnsafeAppend(index);
          }
          ++index;
        },
        [&]() {
          ++index;
        });
    return Status::OK();
  }
```

Reply:
Good catch, it's a leftover from the previous implementation where this block was invoking `Resize` and thus throwing a possible error.